File: old/2020/18_operation_order/aoc18.jl
# Operation Order
#
# Advent of Code 2020, day 18
# Solution by Geir Arne Hjelle, 2020-12-18

module AOC18

using Pipe


function add_to_value(value, op, number)
    if op == '+'
        value + parse(Int, number)
    elseif op == '*'
        value * parse(Int, number)
    end
end

function parse_left_right(expression)
    value, op, number = 0, '+', ""
    for char in expression
        if char ∈ ['+', '*']
            value = add_to_value(value, op, number)
            number = ""
            op = char
        elseif char ∈ "0123456789"
            number = number * char
        end
    end
    add_to_value(value, op, number)
end

function parse_plusses_first(expression)
    while true
        m = match(r"((\d+) [+] (\d+))", expression)
        if m === nothing
            break
        end
        value = parse(Int, m[2]) + parse(Int, m[3])
        expression = replace(expression, m[1] => string(value), count=1)
    end
    parse_left_right(expression)
end

function parse_parens(calculate)
    function _parse_parens(line)
        while true
            # Match unnested parentheses
            m = match(r"(\([^)(]*?\))", line)
            if m === nothing
                break
            end
            value = calculate(m[1][2:end - 1])
            line = replace(line, m[1] => string(value))
        end
        calculate(line)
    end
    _parse_parens
end

# Solve the problem for one file
function solve(input)
    # Parse input
    expressions = split(input, "\n")

    # Part 1
    part_1 = expressions .|> parse_parens(parse_left_right) |> sum

    # Part 2
    part_2 = expressions .|> parse_parens(parse_plusses_first) |> sum

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module

File: old/2016/18_like_a_rogue/aoc18.py
"""Like a Rogue

Advent of Code 2016, day 18
Solution by Geir Arne Hjelle, 2017-05-28
"""

# Standard library imports
import pathlib
import sys

TRAPS = {"^^.", ".^^", "^..", "..^"}


def next_row(row):
    new_tiles = list()
    for tiles in zip("." + row, row, row[1:] + "."):
        new_tiles.append("^" if "".join(tiles) in TRAPS else ".")

    return "".join(new_tiles)


def map_room(first_row, num_rows):
    rows = [first_row]
    while len(rows) < num_rows:
        rows.append(next_row(rows[-1]))

    if "--draw" in sys.argv and num_rows < 1000:
        print("\n".join(rows))

    return rows


def main(args):
    """Solve the problem for all file paths"""
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        solve(file_path)


def solve(file_path):
    """Solve the problem for one file path"""
    print(f"\n{file_path}:")
    with file_path.open(mode="r") as fid:
        for line in fid:
            num_rows, first_row = line.strip().split()
            rows = map_room(first_row, int(num_rows))
            num_safe = sum(r.count(".") for r in rows)
            print(f"{num_safe} safe tiles in {num_rows} rows")


if __name__ == "__main__":
    main(sys.argv[1:])
